lib/code-caller-docker.js

```diff
@@ -0,0 +1,743 @@
+    if (!this._checkState([CREATED, WAITING, EXITING, EXITED])) {
+      callback(new Error('Unexpected DockerCaller state'));
+      return;
+    }
+
+    if (this.callCount === 0) {
+      // If there weren't any calls since the last time this code caller
+      // was restarted, we can slightly optimize things by skipping the
+      // restart. This is safe, as no user-provided code will have been
+      // loaded into the Python interpreter.
+      this.debug(`exit restart() - skipping since no calls recorded since last restart`);
+      callback(null, true);
+    } else if (this.state === CREATED) {
+      // no need to restart if we don't have a worker
+      this.debug(`exit restart()`);
+      callback(null, true);
+    } else if (this.state === WAITING) {
+      this.call('restart', null, null, 'restart', [], (err, ret_val, _consoleLog) => {
+        this.coursePath = null;
+        this.callCount = 0;
+        if (ERR(err, callback)) return;
+        if (ret_val !== 'success') return callback(new Error(`Error while restarting: ${ret_val}`));
+        this.debug('exit restart()');
+        callback(null, true);
+      });
+    } else if (this.state === EXITING || this.state === EXITED) {
+      this.debug('exit restart()');
+      callback(null, false);
+    }
+  }
+
+  done() {
+    this.debug('enter done()');
+    this._checkState([CREATED, WAITING, EXITING, EXITED]);
+
+    if (this.state === CREATED) {
+      this.state = EXITED;
+    } else if (this.state === WAITING) {
+      this._cleanup();
+      this.state = EXITING;
+    }
+    this._checkState();
+    this.debug('exit done()');
+  }
+
+  async ensureChild() {
+    this.debug('enter ensureChild()');
+    this._checkState();
+
+    // This will only execute if we're in `container` mode.
+    // Since container creation is async, it's possible that ensureChild()
+    // could be called again while it's already executing. For instance, we
+    // could be inside a call that was made to warm up this worker, but then
+    // we might have call() invoked, which will also call ensureChild(). We
+    // need to ensure that we're only ever inside this function once at a time,
+    // so we'll use a "mutex" (even though we're in a single-threaded environment).
+    await this.ensureChildMutex.runExclusive(async () => {
+      if (this.container) {
+        this.debug('exit ensureChild() - existing container');
+        return;
+      }
+      await ensureImage();
+      await this._ensureDirectoryIfNeeded();
+      await this._createAndAttachContainer();
+      this.state = WAITING;
+      this._checkState();
+      this.debug('exit ensureChild()');
+    });
+  }
+
+  async _ensureDirectoryIfNeeded() {
+    await this.ensureDirectoryMutex.runExclusive(async () => {
+      if (config.workersExecutionMode === 'container' && !this.hostDirectory) {
+        this.hostDirectory = await tmp.dir({ unsafeCleanup: true, prefix: MOUNT_DIRECTORY_PREFIX });
+      }
```

Comment:
This can be simplified - `code-caller-docker` is only ever used when `workersExecutionMode === 'container'`, so this condition will always be true.
```suggestion
      if (this.hostDirectory) return;
      this.hostDirectory = await tmp.dir({ unsafeCleanup: true, prefix: MOUNT_DIRECTORY_PREFIX });
```